src/main/scala/devices/tilelink/CLINT.scala

```diff
@@ -12,108 +12,53 @@ import freechips.rocketchip.subsystem._
-
-  val intnode : IntNexusNode = IntNexusNode(
-    sourceFn = { _ => IntSourcePortParameters(Seq(IntSourceParameters(ints, Seq(Resource(device, "int"))))) },
-    sinkFn   = { _ => IntSinkPortParameters(Seq(IntSinkParameters())) },
-    outputRequiresInput = false)
-
-  lazy val module = new Impl
-  class Impl extends LazyModuleImp(this) {
-    Annotated.params(this, params)
-    require (intnode.edges.in.size == 0, "CLINT only produces interrupts; it does not accept them")
-
-    val io = IO(new Bundle {
-      val rtcTick = Input(Bool())
-    })
-
-    val time = RegInit(0.U(timeWidth.W))
-    when (io.rtcTick) { time := time + 1.U }
-
-    val nTiles = intnode.out.size
-    val timecmp = Seq.fill(nTiles) { Reg(UInt(timeWidth.W)) }
-    val ipi = Seq.fill(nTiles) { RegInit(0.U(1.W)) }
-
-    val (intnode_out, _) = intnode.out.unzip
-    intnode_out.zipWithIndex.foreach { case (int, i) =>
-      int(0) := ShiftRegister(ipi(i)(0), params.intStages) // msip
-      int(1) := ShiftRegister(time.asUInt >= timecmp(i).asUInt, params.intStages) // mtip
-    }
-
-    /* 0000 msip hart 0
-     * 0004 msip hart 1
-     * 4000 mtimecmp hart 0 lo
-     * 4004 mtimecmp hart 0 hi
-     * 4008 mtimecmp hart 1 lo
-     * 400c mtimecmp hart 1 hi
-     * bff8 mtime lo
-     * bffc mtime hi
-     */
-
-    node.regmap(
-      0                -> RegFieldGroup ("msip", Some("MSIP Bits"), ipi.zipWithIndex.flatMap{ case (r, i) =>
-        RegField(1, r, RegFieldDesc(s"msip_$i", s"MSIP bit for Hart $i", reset=Some(0))) :: RegField(ipiWidth - 1) :: Nil }),
-      timecmpOffset(0) -> timecmp.zipWithIndex.flatMap{ case (t, i) => RegFieldGroup(s"mtimecmp_$i", Some(s"MTIMECMP for hart $i"),
-          RegField.bytes(t, Some(RegFieldDesc(s"mtimecmp_$i", "", reset=None))))},
-      timeOffset       -> RegFieldGroup("mtime", Some("Timer Register"),
-        RegField.bytes(time, Some(RegFieldDesc("mtime", "", reset=Some(0), volatile=true))))
-    )
-  }
-}
+case object CLINTKey extends Field[Option[CLINTParams]](None)
 
-/** Trait that will connect a CLINT to a subsystem */
+// if is isACLINT is false, a clint will be generated according to base address of mswi device
 trait CanHavePeripheryCLINT { this: BaseSubsystem =>
-  val clintOpt = p(CLINTKey).map { params =>
-    val tlbus = locateTLBusWrapper(p(CLINTAttachKey).slaveWhere)
-    val clint = LazyModule(new CLINT(params, cbus.beatBytes))
-    clint.node := tlbus.coupleTo("clint") { TLFragmenter(tlbus) := _ }
-
-    // Override the implicit clock and reset -- could instead include a clockNode in the clint, and make it a RawModuleImp?
-    InModuleBody {
-      clint.module.clock := tlbus.module.clock
-      clint.module.reset := tlbus.module.reset
+  val mswiOpt = p(CLINTKey) map { clintParams =>
+    val mswiOpt = clintParams.mswi.map { params =>
+      val tlbus = locateTLBusWrapper(p(MSWIAttachKey).slaveWhere)
+      val mswi = LazyModule(new MSWI(params, MTIMERParams(), clintParams.isACLINT, cbus.beatBytes))
```

Comment:
shouldn't this be clintParams.mtimer?

Reply:
Also tlbus.beatBytes

Reply:
ClintParams.mtimer is indeed better. But I think it's `cbus. beatBytes`, as clint is cbus device which is consistent with the previous version.

Reply:
The attach point of the clint is parameterized to the `slaveWhere` of `MSWIAttachKey`, so it should use that bus

Reply:
That's true, but I'm consistent with the previous version of CLINT. I think sequencer believes that CLINT will only be available on CBUS.